internal/pkg/earpointer/taint.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package earpointer
+
+import (
+	"go/types"
+
+	"github.com/google/go-flow-levee/internal/pkg/source"
+	"golang.org/x/tools/go/ssa"
+)
+
+// A data structure to assist the traversal of an EAR heap.
+type heapTraversal struct {
+	partitions *Partitions
+	callees    map[*ssa.Function]bool // the functions containing the references of interest
+	visited    ReferenceSet           // the visited references during the traversal
+}
+
+// For a function, transitively get the functions called within this function.
+// Argument "depth" controls the depth of the call chain.
+// For example, return {g1,g2,g3} for "func f(){ g1(); g2() }, func g1(){ g3() }".
+func getCalleeFunctions(fn *ssa.Function, result map[*ssa.Function]bool, depth uint) {
```

Comment:
What is the evidence we need a depth parameter?

Reply:
As mentioned before, this is to control how far we go when searching the call chain. In the future CLs it will be used to control the searching of other cases. For example, one interesting case is to chase the chain of callers: f2->g2->h2.

```
func f2() string {
   a := new T{} //  taint source
   return fmt.Printf(a)
}

func g2() {
   return f2(a)
}

func h2() { 
   x := g2()
  sinkf(x) // taint sink
}
```

Another interesting case is:
```
func f3() string {
   a := new T{} //  taint source
   return fmt.Printf(a)
}

func g3(x string) {
     sinkf(x) // taint sink
}

func h3() { 
   x := f3()
   g3(x)
}
```

So connecting a source and a sink may span many functions. This `depth` is used to limit how many functions we should consider when connecting a source and a sink. Ideally we should not use such a limit, but since EAR uses aggressive merging, the FP rate may be high without applying this limit. Too many references may be involved as well if we don't bound it. If later on we have experimental data justifying that we don't need such a limit, then we can remove it.
I recall that HawkEye uses a similar limit.

Reply:
So the evidence at the moment is 1) theoretical 2) following practices from similar projects. Is this a fair summary? These are not terrible reasons, but they are not overwhelming either.

> If later on we have experimental data justifying that we don't need such a limit, then we can remove it.

It is easy to go from no limit to adding one too.

Reply:
I believed that I've done some tests by disabling the limit or setting a large limit, there were more FPs showing up. 
Anyway, this can be gotten around by using a large limit so I don't think that it is a fundamental issue.